Add min/max range of the event.ingested field to cluster state for searchable snapshots

quux00 opened the pull request:
Add `event.ingested` min/max range to cluster state for searchable snapshots in order
to support shard skipping on the search coordinator node for event.ingested
like we do for the `@timestamp` field.

The key motivation is that the Elastic Security solution uses two timestamp fields:
`@timestamp` which is user populated and `event.ingested` which is populated by
an ingest processor which tells the time that the event was ingested.

In some cases, queries are filtered by @timestamp, while in some others they are filtered
by event.ingested. When data is stored in the cold or frozen tier, we have a shard skipping
mechanism on the coordinating node, which allows the coordinating node to skip shards
based on min and max value stored in the cluster state for the @timestamp field. This is
done to prevent returning irrelevant errors in case shards in frozen/cold are unavailable
(as they have no replicas) and queries that don’t target them (as the datetime filter won’t
match them) are executed.

This works well for the @timestamp field and the Security solution could benefit from having
the same skipping mechanism applied to the event.ingested field.

quux00 committed:
Add event.ingested min/max to cluster state for searchable snapshots

quux00 committed:
Added DateFieldRange to CoordinatorRewriteContext and field names to each method on that class

quux00 committed:
Changed mappingSupplier in CoordinatorRewriteContextProvider to take Map<String, DateFieldMapper.DateFieldType>>

quux00 committed:
ShardStateAction now supports eventIngestedRange (no new tests however)

ShardStateAction.execute updates eventIngestedRange
and StartedShardEntry now has ShardLongFieldRange eventIngestedRange
but it is not yet populated (set to UNKNOWN) - addressed in next commit

quux00 committed:
Added eventIngestedRange to ShardStateAction constructor and all knock-on changes related.

No testing against eventIngestedRange yet.
Also see now that I need to add eventIngested support to RecoveryListener.onRecoveryDone intf

quux00 committed:
Added eventIngestedMillisFieldRange to onRecoveryDone

Next need to implement getEventIngestedRange on concrete impls of
IndicesClusterStateService.Shard interface

quux00 committed:
Implement getEventIngestedRange on concrete impls of IndicesClusterStateService.Shard interface

quux00 commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContextProvider.java:
```diff
@@ -49,18 +51,26 @@ public CoordinatorRewriteContext getCoordinatorRewriteContext(Index index) {
         IndexLongFieldRange timestampRange = indexMetadata.getTimestampRange();
-        if (timestampRange.containsAllShardRanges() == false) {
-            timestampRange = indexMetadata.getTimeSeriesTimestampRange(dateFieldType);
+        if (timestampRange.containsAllShardRanges() == false) { /// MP TODO: is this logic needed for event.ingested? Not clear what this is
```
@martijnvg - Could you help me understand what `timestampRange.containsAllShardRanges()` is doing and why it being false means we need to check `getTimeSeriesTimestampRange()`?

And does that check have any bearing on the event.ingested timestamp range?

martijnvg replied:
It has been a while since I looked at this. But the background is that for the `@timestamp` we do additional filtering of shards. 

Each tsdb backing index has an `index.time_series.start_time` and `index.time_series.end_time` index settings. Each document has `@timestamp` field. During indexing documents are routed to the backing index that matches with the start end and time range.

For searches with a range query on `@timestamp` field, this gives us the ability to filter out shards that have backing indices that do not match with the range.

This logic here is a fallback to use the time serie range (based on start and end time index settings in cluster state) in case the range matches with shards that are not read only (I guess not yet frozen or backed by searchable snapshot). Obviously this only applies to tsdb and the `@timestamp` field. So I think this doesn't apply to the `event.ingested` field.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Added IT tests for event.ingested. SearchableSnapshot based tests are failing

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Uncommented failing line in SearchableSnapshotsCanMatchOnCoordinatorIntegTests

quux00 commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsCanMatchOnCoordinatorIntegTests.java:
```diff
@@ -250,19 +255,35 @@ public void testSearchableSnapshotShardsAreSkippedBySearchRequestWithoutQuerying
         assertThat(dateFieldType, notNullValue());
         final DateFieldMapper.Resolution resolution = dateFieldType.resolution();
         assertThat(updatedTimestampMillisRange.isComplete(), equalTo(true));
+        assertThat(updatedEventIngestedMillisRange.isComplete(), equalTo(true)); /// MP TODO: this sometimes fails - why?
```
The assert on `updatedTimestampMillisRange.isComplete()` never fails.

Out of ~30 runs I've done of this test, the `updatedEventIngestedMillisRange.isComplete()` fails roughly 80% of the time. I tried to find any correlations of what it fails with (such as the several randomized factors in this test), but was unable to find any correlation. I also tried adding a 10s sleep to see if there was some race condition that need time, but it fails sporadically even with the sleep.

So somewhere there is proper handling of the `@timestamp` IndexLongFieldRange to provide info on all shards that is not happening for the newly added `event.ingested` range, but I'm not able to figure out what/where that is.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Added missing this.eventIngestedRange = indexMetadata.eventIngestedRange; to IndexMetadata.Builder

Removed null checks in IndexMetadata
Added eventIngestedRangeInfo to AbstractBuilderTestCase

quux00 committed:
Fixed logic error in SearchableSnapshotsCanMatchOnCoordinatorIntegTests

Added reading/writing of event.ingested in serialization code of static IndexMetadata.readFrom method and writeTo method

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Changed from hardcoded "event.ingested" to IndexMetadata.EVENT_INGESTED_FIELD_NAME

Updated javadoc in CoordinatorRewriteContext

quux00 committed:
Updated javadoc. Improved SearchableSnapshotsCanMatchOnCoordinatorIntegTests.

One test seed that consistently fails with this code:
./gradlew -p x-pack/plugin/frozen-indices internalClusterTest --tests "org.elasticsearch.index.engine.frozen.FrozenIndexIT.testTimestampFieldTypeExposedByAllIndicesServices" -Dtests.seed=93869665C2B3B240

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -756,20 +806,27 @@ public static class StartedShardEntry extends TransportRequest {
+            if (in.getTransportVersion().onOrAfter(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
+                this.eventIngestedRange = ShardLongFieldRange.readFrom(in);
+            } else {
+                this.eventIngestedRange = ShardLongFieldRange.UNKNOWN;  // MP TODO: is this the right choice?
```
Yes, `UNKNOWN` makes sense for bwc.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -651,32 +673,56 @@ public ClusterState execute(BatchExecutionContext<StartedShardUpdateTask> batchE
                             );
-                            if (newTimestampMillisRange != currentTimestampMillisRange) {
-                                updatedTimestampRanges.put(index, newTimestampMillisRange);
+                            final IndexLongFieldRange newEventIngestedMillisRange = currentEventIngestedMillisRange.extendWithShardRange(
```
I think we should only track this if `batchExecutionContext.initialState().getMinTransportVersion().onOrAfter(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)`. Otherwise it should be `UNKNOWN`. If we don't do that, we will have different cluster states on different nodes because we can't send this data over the wire to older nodes.

quux00 replied:
Changed in next push.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetadataUpdater.java:
```diff
@@ -169,8 +169,10 @@ private static IndexMetadata updateInSyncAllocations(
-                    updatedIndexMetadata = updatedIndexMetadata.withTimestampRange(
-                        updatedIndexMetadata.getTimestampRange().removeShard(shardId.id(), oldIndexMetadata.getNumberOfShards())
+                    updatedIndexMetadata = updatedIndexMetadata.withTimestampRanges(
+                        /// MP TODO: Is it safe/OK to call removeShard here twice? Not sure why we call removeShard on this to start with.
```
Yes, I think this is needed. We're removing the shard from the two different ranges because we're creating it again from a stale copy that might in principle have different data, which will be filled in again on the subsequent shard-started event.

quux00 committed:
Changes to ShardStateAction, based on PR feedback.

ShardStartedClusterStateTaskExecutor.execute now checks min
transport version against TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE
before assigning a value to the event.ingested range. If min version is not met, then
event.ingested range is set to UNKNOWN.

quux00 committed:
event.ingested added to from/toXContent of IndexMetadata

Unfortunately this now causes tests that were passing to fail. Proper solution for this not yet known.

quux00 committed:
Removed TODO

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Hardcoded index=true in SearchableSnapshotsIntegTests to get help debugging why the test fails

quux00 committed:
Fixed minor assert bug in ShardStateAction around event.ingested. Updated javadocs in ShardStateAction

quux00 committed:
Added eventIngestedRange to RestoreService

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Fixed ClusterStateTests

quux00 committed:
Fixed typo in ClusterStateTests

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Fixed additional tests

quux00 committed:
Fixed ShardStartedClusterStateTaskExecutorTests to pass

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Fixed failing FrozenIndexIT test by adding event.ingested handling to MetadataIndexStateService, which was missed in earlier passes

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Added new test in FrozenIndexIT

quux00 committed:
Added event.ingested to ShardStartedClusterStateTaskExecutorTests; general clean up for initial review

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 committed:
Update docs/changelog/106252.yaml

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

quux00 committed:
Fixed changelog added by GitHub

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Removed assert in ShardStateAction that requires event.ingested to have a complete timestamp range

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -725,6 +786,14 @@ private static boolean assertStartedIndicesHaveCompleteTimestampRanges(ClusterSt
+                // + "] should have complete event.ingested range, but got "
+                // + clusterState.metadata().index(cursor.getKey()).getEventIngestedRange()
+                // + " for "
+                // + cursor.getValue().prettyPrint();
```
This assert causes many of the bwc tests to fail. I've commented it out here rather than delete it, to see if anyone thinks that some form this assert should be used (and suggestions on how to do it in a backwards compatible way).

DaveCTurner replied:
Hmm I think if this assert doesn't hold then there's a bug somewhere. In a BwC situation the range should be `UNKNOWN` which is indeed complete.

quux00 replied:
Got it. Thanks David.

I added some logging for when this assert fails and the event.ingested range is coming in as `NO_SHARDS` for the tests that fail.

quux00 replied:
Since existing cluster state (of older clusters) doesn't include event.ingested range, should we instead default eventIngestedRange to UNKNOWN rather than NO_SHARDS?

I tried there: 
https://github.com/elastic/elasticsearch/pull/106252/commits/3bd5830613fa482f8d985f98e32cc7a719b830a9#diff-d9e713c7c9f70d3b97888b7ed019ba4469397004333b0d0c13fa98442beef047R1882

and with that commit all the bwc tests now pass (as well as non-bwc tests after I adjusted a few).

DaveCTurner replied:
Hmm I think we want to default to `NO_SHARDS` here. Instead should we treat metadata received from older nodes as if it were `UNKNOWN`?

```diff
diff --git a/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java b/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java
index 0e92b3acf12..8afee04136b 100644
--- a/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java
+++ b/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java
@@ -1780,6 +1780,8 @@ public class IndexMetadata implements Diffable<IndexMetadata>, ToXContentFragmen
         }
         if (in.getTransportVersion().onOrAfter(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
             builder.eventIngestedRange(IndexLongFieldRange.readFrom(in));
+        } else {
+            builder.eventIngestedRange(IndexLongFieldRange.UNKNOWN);
         }
         return builder.build(true);
     }
@@ -1832,6 +1834,8 @@ public class IndexMetadata implements Diffable<IndexMetadata>, ToXContentFragmen
         }
         if (out.getTransportVersion().onOrAfter(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
             eventIngestedRange.writeTo(out);
+        } else {
+            assert eventIngestedRange == IndexLongFieldRange.UNKNOWN : eventIngestedRange;
         }
     }
```

Similarly for `IndexMetadataDiff`.

I think we should also have an assertion somewhere that all these values are `UNKNOWN` until all nodes in the cluster support the new feature.

quux00 committed:
Added logging around assert in ShardStateAction that causes bwc errors to help debug cause of test failures

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Changed assert to include IndexLongFieldRange.UNKNOWN to see if that allows bwc tests to pass

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Fixed compile error in SearchableSnapshotsCanMatchOnCoordinatorIntegTests and corrected log message in ShardStateAction for bwc testing

quux00 committed:
Added logging to help debug bwc tests

quux00 committed:
Modified temp logging in ShardStateAction to debug failing bwc tests

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Additional temp logging to determine where NO_SHARDS is coming from for event.ingested in bwc tests

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Additional logging added for debugging bwc tests

quux00 committed:
eventIngestedRange = IndexLongFieldRange.UNKNOWN is default in IndexMetadata to see if that allows bwc tests to pass

quux00 committed:
Setting IndexLongFieldRange.UNKNOWN in IndexMetadata readFrom and asserts for UNKNOWN in writeTo methods

quux00 committed:
Removed 'assert eventIngestedRange == IndexLongFieldRange.UNKNOWN' from IndexMetadata to see if bwc tests pass

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Setting IndexLongFieldRange.UNKNOWN in IndexMetadata.fromXContent as default; removed debugging logging

quux00 committed:
Additional logging to debug failing mixed-cluster tests; Uncommented assert for event.ingested==UNKNOWN

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Updated TransportVersions

quux00 committed:
Stable point for new test testEventIngestedRangeInSearchAgainstSearchableSnapshotShards; needs cleanup

quux00 committed:
Trying to set UNKNOWN for event.ingested in apply of IndexMetadataDiff to see if that works for mixed-cluster tests

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Finalized new test in SearchableSnapshotsCanMatchOnCoordinatorIntegTests for event.ingested; passing

quux00 committed:
Set eventIngestedRange=UNKNOWN in MetadataCreateIndexService if min transport version before EVENT_INGESTED_RANGE_IN_CLUSTER_STATE

Hoping this fixes the index creation test failures in mixed-cluster tests.

Also added new tests for testing the new logic in CoordinatorRewriteContextProvider.getCoordinatorRewriteContext
which was changed in previous commit

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Fix checkstyle issue

quux00 committed:
Added min transport version check to MetadataIndexStateService

quux00 committed:
Added min transport version check to RestoreService

quux00 committed:
Added min transport version check for setting UNKNOWN in MetadataIndexStateService

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Finished new tests in CanMatchPreFilterSearchPhaseTests

quux00 committed:
Cleanup and some minor test tweaks to ready this for review

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 commented:
@elasticmachine run elasticsearch-ci/part-2

quux00 commented:
@elasticmachine run packaging-tests-unix-sample

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Reverted 10 files back to main in order to focus this PR on just changes to cluster state (adding event.ingested range).

Changes reverted were around the search can-match Coordinator rewrite logic using the new event.ingested range
and associated tests. Those will be added in a follow-on PR.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -2298,6 +2354,7 @@ IndexMetadata build(boolean repair) {
                 isSystem,
                 INDEX_HIDDEN_SETTING.get(settings),
                 timestampRange,
+                eventIngestedRange,
```
Can we assert that it is `!= null` in the `build` method?

quux00 replied:
Added in next push.

henningandersen commented on server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:
```diff
@@ -1742,6 +1747,8 @@ private static IndexMetadata.Builder restoreOverClosedIndex(IndexMetadata snapsh
             .settingsVersion(Math.max(snapshotIndexMetadata.getSettingsVersion(), 1 + currentIndexMetadata.getSettingsVersion()))
             .aliasesVersion(Math.max(snapshotIndexMetadata.getAliasesVersion(), 1 + currentIndexMetadata.getAliasesVersion()))
             .timestampRange(IndexLongFieldRange.NO_SHARDS)
+            // TODO: we don't have access to cluster state here, so we can't check min transport version - will that be a problem?
```
I think it will, since there are assertions validating that we are not sending this state to an older node. Seems we have the cluster state one stack frame out, so seems easily solvable?

quux00 replied:
Good idea. Fixed in next push.

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java:
```diff
@@ -578,7 +592,8 @@ public void testStartedShardEntrySerialization() throws Exception {
 
         final TransportVersion version = randomFrom(randomCompatibleVersion(random()));
```
Can we add a serialization test verifying that we can send the entry and receive the entry to/from an older version too?

quux00 replied:
New test added here: https://github.com/elastic/elasticsearch/pull/106252/commits/ccd11d95f16055c11fdc9ca2a3fd8a8f0b6c87d1#diff-e8639cf8d716c3052fc07cc779e4a229595fc7292d31939c2e1130ab6b4ae4aaR611

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexServiceTests.java:
```diff
@@ -1096,7 +1097,16 @@ public void testBuildIndexMetadata() {
+            4,
+            sourceIndexMetadata,
+            false,
+            TransportVersion.current()
```
Can we also test with an older minimum version? As well as add assertions on the expected eventIngested range?

quux00 replied:
New test added in this commit: https://github.com/elastic/elasticsearch/pull/106252/commits/282170de4c5943c3bed71029ead70fccb4609f36

Let me know if that's not what you had in mind.

henningandersen replied:
That looks right to me, but I'd like to have an assertion in the test on the expected event.ingest range, which I think should be `UNKNOWN` for older transport versions and `NO_SHARDS` for newer?

quux00 replied:
Yes. I will fix that in next push.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:
```diff
@@ -723,32 +740,54 @@ public void testSnapshotMountedIndexWithTimestampsRecordsTimestampRangeInIndexMe
             .clear()
             .setMetadata(true)
             .setIndices(indexName)
             .get()
             .getState()
             .metadata()
-            .index(indexName)
-            .getTimestampRange();
+            .index(indexName);
 
+        final IndexLongFieldRange timestampRange = indexMetadata.getTimestampRange();
         assertTrue(timestampRange.isComplete());
 
+        final IndexLongFieldRange eventIngestedRange = indexMetadata.getEventIngestedRange();
+        assertTrue(eventIngestedRange.isComplete());
+
         if (indexed) {
             assertThat(timestampRange, not(sameInstance(IndexLongFieldRange.UNKNOWN)));
+            assertThat(eventIngestedRange, not(sameInstance(IndexLongFieldRange.UNKNOWN)));
             if (docCount == 0) {
                 assertThat(timestampRange, sameInstance(IndexLongFieldRange.EMPTY));
+                assertThat(eventIngestedRange, sameInstance(IndexLongFieldRange.EMPTY));
             } else {
                 assertThat(timestampRange, not(sameInstance(IndexLongFieldRange.EMPTY)));
-                DateFieldMapper.Resolution resolution = dateType.equals("date")
+                DateFieldMapper.Resolution timestampResolution = dateType.equals("date")
```
This rename seems unnecessary and I think we can use the same `resolution` for validating both timestamp ranges, since we use the same `dateType` anyway?

quux00 replied:
I don't think so? My understanding is that resolution can be different between the two - one could be millis resolution and the other could be nanos. Or is there some invariant enforced in the production code somewhere that both @timestamp and `event.ingested` have to use the same resolution?

henningandersen replied:
It is more to do with the test code. [Here](https://github.com/elastic/elasticsearch/pull/106252/files/4dcfea5dfb5dda3e6af760a957f7975570833c50#diff-f1bb44f1e1546df7098770be6ad7127aea7fa2ff3e7d4601e90e4d3ad520944fR673) we pick a single one of ("date", "date_nanos"), used for both fields in this test. Thus the expected resolution of both fields will also be the same due to how the test is written.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -1609,6 +1634,11 @@ private static class IndexMetadataDiff implements Diff<IndexMetadata> {
                 indexWriteLoadForecast = null;
                 shardSizeInBytesForecast = null;
             }
+            if (in.getTransportVersion().onOrAfter(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
+                eventIngestedRange = IndexLongFieldRange.readFrom(in);
+            } else {
+                eventIngestedRange = IndexLongFieldRange.UNKNOWN;
+            }
```
Can we add serialization test for index metadata that ensures we properly serialize the event ingested field - and also that bwc serialization works?

quux00 replied:
I have a first pass at that in this new commit: https://github.com/elastic/elasticsearch/pull/106252/commits/5fb26b88e65792f79f7977ed4e2abf784060b0f6

I'm not clear on how exactly to structure a bwc test. I can say that earlier in the PR, several existing bwc (and mixed-cluster) tests were failing for several days while I tracked down all the places to set the eventIngested to UNKNOWN in order to get them passing again. I'd welcome any guidance on how to structure a unit test "bwc" test that focuses on this code.

henningandersen replied:
Revisiting this, I think fixing the todo you added in `IndexMetadataTests` should be enough. In order to do bwc testing, we'd have to manually serialize the indexmetadata object, which seems unnecessary given that we should be certain that any rolling upgrade tests will do so.

quux00 replied:
Sounds good. New version of that test in next push.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -2522,6 +2584,9 @@ public static IndexMetadata fromXContent(XContentParser parser, Map<String, Mapp
                         case KEY_TIMESTAMP_RANGE:
                             builder.timestampRange(IndexLongFieldRange.fromXContent(parser));
                             break;
+                        case KEY_EVENT_INGESTED_RANGE:
```
Can we add a test for from xcontent that verifies that we get the right event ingested field and also that the bwc works when it is missing?

quux00 replied:
Thanks.

I've added a simple test for event.ingested in `fromXContent` here: https://github.com/elastic/elasticsearch/pull/106252/commits/5b1e47bb41cba19b0632acd9ac7596511074457a

There are dozens of tests that rely on this behavior now (if you comment out this code lots of tests, including IT tests fail).

And for bwc, like my previous comment, I had to do various commits to get existing bwc tests passing with earlier versions of this PR. How would you recommend structuring a bwc test for this code? With XContent (vs. Writeables) I'm even less clear how to structure a bwc test here, so I welcome input on how to set that up (for instance, is there a known good example of doing this you can point at)?

henningandersen replied:
I would write the xcontent (like what is done in `IndexMetadataTests.testIndexMetadataSerialization`), strip the field out and then check that `fromXContent` can parse it. I prefer to have an explicit test for things like this, since it is very serious if this should not work (and the rolling upgrade tests etc that caught the issue so far may change in the future).

quux00 replied:
Got it. Thanks. I have made an attempt at that in the [latest commit](https://github.com/elastic/elasticsearch/pull/106252/commits/56e4d1a384735c9017f8ae0daaa51baea4710c2f).

henningandersen replied:
:+1: looks good

henningandersen reviewed (commented):
Sorry for the delay here. Looks good. I left a number of smaller comments, except perhaps the one about the default value of event ingested range.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -1843,6 +1891,7 @@ public static class Builder {
         private Integer routingNumShards;
         private boolean isSystem;
         private IndexLongFieldRange timestampRange = IndexLongFieldRange.NO_SHARDS;
+        private IndexLongFieldRange eventIngestedRange = IndexLongFieldRange.NO_SHARDS;
```
I tried doing that early on in this PR when bwc tests started failing: https://github.com/elastic/elasticsearch/pull/106252#discussion_r1592835564

However, @DaveCTurner argued instead that we need to default to NO_SHARDS and assign (and assert for) UNKNOWN in the mixed version scenarios.

> it is relatively hard to reason about having caught all places with the current default

Very true. That's why it took me a long time to track down all the places that bwc and mixed cluster tests were failing (it was/is quite opaque to me, since I don't understand all of the close/restore/open/create/recover etc pathways that converge on this). I'm still not convinced I haven't missed any edge cases.

Because of that I'm relying on es-distributed to determine what's the right solution here and what the edge cases area.

henningandersen replied:
From that comment:
> I think we should also have an assertion somewhere that all these values are UNKNOWN until all nodes in the cluster support the new feature.

Did we add such an assertion? I suppose we can add that when building cluster state maybe? That would be just as good I think.

I think getting to a NO_SHARDS default is right, once we remove BWC logic. If adding the above assertion is infeasible, I'd probably prefer to make it null default here and assert that it has an explicit value set (which could be from the prior index metadata definition).

quux00 replied:
> Did we add such an assertion? I suppose we can add that when building cluster state maybe? That would be just as good I think.

Asserts for UNKNOWN are done in two places:

https://github.com/elastic/elasticsearch/pull/106252/files#diff-d9e713c7c9f70d3b97888b7ed019ba4469397004333b0d0c13fa98442beef047R1693

https://github.com/elastic/elasticsearch/pull/106252/files#diff-d9e713c7c9f70d3b97888b7ed019ba4469397004333b0d0c13fa98442beef047R1855

Both of which are in `writeTo` methods. It is not clear to me where else to add them.

And with the [first new commit](https://github.com/elastic/elasticsearch/commit/0e69ac3c7785499cbf3de6977d2c5f414aaebe9b) we now pass in minClusterTransportVersion to IndexMetadata.withTimestampRanges to ensure UNKNOWN is set when there is still a mixed-cluster.

> I suppose we can add that when building cluster state maybe? That would be just as good I think.

[The second new commit](https://github.com/elastic/elasticsearch/pull/106252/commits/11dca2acb2b313735dad45a0b322ebcabdc78842) sort of addresses that. The problem is that `IndexMetadata.build` and `IndexMetadataDiff.build` are called from a lot of places, some of which don't have access to the min cluster transport version. Example: `fromXContent` methods. That would be needed in order to do asserts.

So instead I tried to require passing in `TransportVersion minClusterTransportVersion` into the `Builder#eventIngestedRange`. In many cases, the calling function has access to the minClusterTransportVersion but in a few places (again like fromXContent), they do not, so I pass in `null` in those cases, making this only a partial solution.

Let me know if you think if that is good enough or we need something more strict. Or maybe I'm missing a better way to do it.

> I think getting to a NO_SHARDS default is right, once we remove BWC logic. If adding the above assertion is infeasible, I'd probably prefer to make it null default here and assert that it has an explicit value set (which could be from the prior index metadata definition).

I tried setting the default to the `null`. That causes over 5,200 unit tests to fail due to [this assert](https://github.com/elastic/elasticsearch/pull/106252/files#diff-d9e713c7c9f70d3b97888b7ed019ba4469397004333b0d0c13fa98442beef047R2340) which I added a while back. So I don't think's feasible?

henningandersen replied:
About cluster state building, I imagined just doing this in `ClusterState.Builder.build` or in the ClusterState constructor. It can simply loop over all the indices checking that the event ingested range is UNKNOWN. No validation is needed once the min transport version is above a certain version number. I think this is easily done, so I'd like to have it even if we strengthened it elsewhere.

quux00 replied:
Oh, good idea. Added in next push.

quux00 replied:
Hmm, I still hit a problem with this approach.

First I tried adding the assert (see below) to the second ClusterState constructor. That causes >900 unit tests to fail with errors like:

```
java.lang.AssertionError: event.ingested range should be UNKNOWN but is NO_SHARDS for index: [idx] minTransportVersion: 7170099
```

I then tried adding to the `ClusterState` builder. Same problem (and I have to compute the min version rather than get it for free).

I'm not sure if all these tests failing is indicative of a major problem with the production code or just an artifact of how unit tests are constructed. I ran three IT tests that I know interact to some degree with my code and they had mixed results

* `FrozenIndexIT` passed
* `SearchableSnapshotsCanMatchOnCoordinatorIntegTests` - failed (same error message as shown above)
* `SearchableSnapshotsIntegTests` - some passed, some failed

Does this make you think we should set the default to UNKNOWN instead?

Here's the code I added (this is the variant for adding it to the builder):

```java
            assert assertEventIngestedIsUnknownInMixedClusters(metadata,
                CompatibilityVersions.minimumVersions(compatibilityVersions.values()));

...

        private boolean assertEventIngestedIsUnknownInMixedClusters(Metadata metadata, CompatibilityVersions compatibilityVersions) {
            if (compatibilityVersions.transportVersion().before(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
                for (IndexMetadata indexMetadata : metadata.indices().values()) {
                    assert indexMetadata.getEventIngestedRange() == IndexLongFieldRange.UNKNOWN
                        : "event.ingested range should be UNKNOWN but is "
                        + indexMetadata.getEventIngestedRange() + " for index: " + indexMetadata.getIndex()
                        + " minTransportVersion: " + previous.getMinTransportVersion();
                }
            }
            return true;
        }

```

henningandersen replied:
I think we should solve this by relaxing it to allow both `NO_SHARDS` and `UNKNOWN`. We should then do the same in the `writeTo` assertion.

henningandersen replied:
Hmm, I take that back (for now at least). We should be able to make this work I think. The main problem seems to be the compatiblity version which is not always initialized. I think following diff solves at least one such case:

```
diff --git a/server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java b/server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java
index 8d72715219e..0d057db3c03 100644
--- a/server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java
+++ b/server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java
@@ -172,9 +172,9 @@ public class TransportClusterStateAction extends TransportMasterNodeReadAction<C
 
         if (request.nodes()) {
             builder.nodes(currentState.nodes());
-            builder.nodeIdsToCompatibilityVersions(getCompatibilityVersions(currentState));
             builder.nodeFeatures(getClusterFeatures(currentState));
         }
+        builder.nodeIdsToCompatibilityVersions(getCompatibilityVersions(currentState));
         if (request.routingTable()) {
             if (request.indices().length > 0) {
                 RoutingTable.Builder routingTableBuilder = RoutingTable.builder();
```

I'll experiement a bit more with some unittests too.

henningandersen replied:
Doing it in the constructor like this:

```
        assert compatibilityVersions.isEmpty() || assertEventIngestedIsUnknownInMixedClusters(metadata, this.minVersions);
    }

    private boolean assertEventIngestedIsUnknownInMixedClusters(Metadata metadata, CompatibilityVersions compatibilityVersions) {
        if (compatibilityVersions.transportVersion().before(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
            for (IndexMetadata indexMetadata : metadata.indices().values()) {
                assert indexMetadata.getEventIngestedRange() == IndexLongFieldRange.UNKNOWN
                    : "event.ingested range should be UNKNOWN but is "
                    + indexMetadata.getEventIngestedRange() + " for index: " + indexMetadata.getIndex()
                    + " minTransportVersion: " + compatibilityVersions.transportVersion();
            }
        }
        return true;
    }
```

should also work I think.

henningandersen replied:
Had to change it to:

```
        assert compatibilityVersions.isEmpty()
            || blocks.hasGlobalBlock(STATE_NOT_RECOVERED_BLOCK)
            || assertEventIngestedIsUnknownInMixedClusters(metadata, this.minVersions);
```

and change `eventTimestampRange` to `UNKNOWN` in `ClusterRerouteResponseTests.createClusterState` and `ClusterStateTests.buildClusterState` (and fixing it in the 3 toXContent tests)

quux00 replied:
Yes, sorry. I missed this set of comments in your previous review. I've confirmed that your suggested assert works against server unit tests (with the slight test adjustments you mentioned) and the handful of IT tests I tried. Will be in next push to see if it passes CI and bwc tests.

quux00 committed:
Init PR feedback changes

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -2464,7 +2525,8 @@ public static IndexMetadata fromXContent(XContentParser parser, Map<String, Mapp
             XContentParserUtils.ensureExpectedToken(XContentParser.Token.FIELD_NAME, parser.currentToken(), parser);
             Builder builder = new Builder(parser.currentName());
-
+            // default to UNKNOWN so that older versions without that range handle it properly
```
Fixed in next push.

quux00 commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -1127,12 +1134,21 @@ private ClusterState openIndices(final Index[] indices, final ClusterState curre
                         .settings(updatedSettings)
-                        .timestampRange(IndexLongFieldRange.NO_SHARDS)
-                        .build();
+                        .timestampRange(IndexLongFieldRange.NO_SHARDS);
+
+                    if (currentState.getMinTransportVersion().before(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
+                        // for versions for we added event.ingested range to cluster state, default to UNKNOWN since it will
+                        // not be handled properly for older clusters in a mixed-cluster env
+                        builder.eventIngestedRange(IndexLongFieldRange.UNKNOWN);
+                    } else {
+                        builder.eventIngestedRange(IndexLongFieldRange.NO_SHARDS);
+                    }
+
+                    IndexMetadata newIndexMetadata = builder.build();
```
Thanks, good call. In the next push, I've added a method called `getEventIngestedDefaultValue(minTransportVersion)` and call it directly in the builder both here and in `closeRoutingTable`.

quux00 commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetadataUpdater.java:
```diff
@@ -167,10 +167,12 @@ private static IndexMetadata updateInSyncAllocations(
-                        updatedIndexMetadata.getTimestampRange().removeShard(shardId.id(), oldIndexMetadata.getNumberOfShards())
+                    updatedIndexMetadata = updatedIndexMetadata.withTimestampRanges(
+                        updatedIndexMetadata.getTimestampRange().removeShard(shardId.id(), oldIndexMetadata.getNumberOfShards()),
+                        updatedIndexMetadata.getEventIngestedRange().removeShard(shardId.id(), oldIndexMetadata.getNumberOfShards())
```
Thanks for flagging this. It's a part of the code that makes me very nervous, as I don't really understand what is happening here.

I did ask David Turner about it in an earlier version of the PR: https://github.com/elastic/elasticsearch/pull/106252/files/126eb4ddf091079fbeba09654409bddac78f822e#r1569022493

You are right that `removeShard` can return NO_SHARDS under either of the following conditions:

```
        if (shards == null && numberOfShards == 1) {
```
or
```
        if (shards != null && shards.length == 1 && shards[0] == shardId) {
```

I don't really understand what is going on when we call removeShard here. The comments indicate that this is checking "if we have been force-initializing an empty primary or a stale primary" and if so should remove a shard from the range. 

> I think this would result in the event ingested timestamp range becoming NO_SHARDS and thus break assertions elsewhere?

So when does reverting to NO_SHARDS break things? What is the right way to handle this? Perhaps if eventIngestedRange is already UNKNOWN, we should have an if-guard here to make sure it doesn't revert to NO_SHARDS?

But then we'd have different handling between @timestamp and event.ingested, as [this test](https://github.com/elastic/elasticsearch/blob/main/server/src/test/java/org/elasticsearch/index/shard/IndexLongFieldRangeTests.java#L104) shows that UNKNOWN can convert to NO_SHARDS under some scenarios.

henningandersen replied:
Yeah, going from UNKNOWN to NO_SHARDS is fine when the minimum transport version check allows it. The problem is the missing check for that, not the code otherwise.

I wonder if we should let `withTimeStampRanges` have a transport version and thus centralize the minimum version check there. That could just disregard the ingested timestamp if the check does not pass. And it would make it more obvious that we have caught all cases of manipulating this correctly.

quux00 replied:
OK, thanks. I've tried that in [this commit](https://github.com/elastic/elasticsearch/commit/0e69ac3c7785499cbf3de6977d2c5f414aaebe9b). It required changing several method signatures along the chain.

henningandersen replied:
I'll leave a PR review comment around this. Overall looks ok but I think we can remove at least one of the signature changes.

quux00 commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -651,32 +675,66 @@ public ClusterState execute(BatchExecutionContext<StartedShardUpdateTask> batchE
+                            IndexLongFieldRange newEventIngestedMillisRange = IndexLongFieldRange.UNKNOWN;
+                            TransportVersion minTransportVersion = batchExecutionContext.initialState().getMinTransportVersion();
+                            if (minTransportVersion.onOrAfter(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
+                                newEventIngestedMillisRange = currentEventIngestedMillisRange.extendWithShardRange(
```
I don't think I understand what you are asking. The newly added code around `event.ingested` here mirrors the way the `@timestamp` range is handled a few lines above, with the one modification that the "newRange" gets initialized to UNKNOWN for cases where the min transport version of the cluster is still indicating that not all clusters have upgraded. If all clusters have upgraded to a version that knows how to deal with event.ingested ranges, then we call `extendWithShardRange` off the `currentEventIngestedMillisRange` (not the "newRange") which we get out of index metadata (just like we do for `@timestamp`).

Sorry if I'm missing your concern - can you elaborate on what you think could go wrong here (and how to compensate for it)?

henningandersen replied:
We will write UNKNOWN into the index metadata of all indices while in mixed cluster. Any subsequent rounds here will remain on UNKNOWN.

It is more related to how to handle existing indices, which I think you want to do a separate PR for. We could in principle maybe just fix it here, but then it will require all shards to start before the ranges are available, which does not seem nice (and may have other pitfalls). So I think we'll want a separate PR to populate all UNKNOWN entries.

So nothing to do here for now.

quux00 committed:
First batch of changes based on PR feedback

Added assert for eventIngestedRange != null in IndexMetadata
Simplified/collated logic for setting eventIngestedRange in MetadataIndexStateService
New test added ShardStateActionTests to test serialization against older versions

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
PR feedback: added test to MetadataCreateIndexServiceTests, building IndexMetadata with older TransportVersions

quux00 committed:
Init versions of two new tests to ClusterSerializationTests

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Added eventIngested to IndexMetadataTests.testIndexMetadataSerialization

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Improved several tests based on PR feedback

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Added minTransportVersion to IndexMetadata.withTimestampRanges in order to keep mixed-clusters with UNKNOWN for event.ingested range

quux00 committed:
Fix ReactiveStorageDeciderServiceTests to compile with new signature for updateMetadataWithRoutingChanges method

quux00 committed:
Changed IndexMetadata.Builder.eventIngestedRange to take in minClusterTransportVersion. Unfortunately not all callers know it.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

henningandersen reviewed (commented):
A few quick initial comments after your latest changes.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
PR changes: Get minTransportVersion from instance var in AllocationService

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:
```diff
@@ -681,21 +681,29 @@ public void testSnapshotMountedIndexLeavesBlobsUntouched() throws Exception {
 
```
The original timestamp PR had a couple assertions more added in this integtest, let us add those for event ingested timestamp too:

https://github.com/elastic/elasticsearch/pull/65689/files#diff-f1bb44f1e1546df7098770be6ad7127aea7fa2ff3e7d4601e90e4d3ad520944fR238

henningandersen replied:
Likewise for `FrozenIndexIT` and `FrozenIndexTests`.

quux00 replied:
Thanks for flagging. Added in next push.

henningandersen reviewed (commented):
This looks good.

I left a number of comments (and would like to have the assertion added as per prior comment thread).

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -912,13 +924,24 @@ public IndexMetadata withIncrementedPrimaryTerm(int shardId) {
             return this;
         }
+        IndexLongFieldRange allowedEventIngestedRange = eventIngestedRange;
+        if (minClusterTransportVersion.before(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
```
OK, but help me understand - what version do you mean would be removed? The `EVENT_INGESTED_RANGE_IN_CLUSTER_STATE` TransportVersion? Or do you mean a specific Elasticsearch version?

quux00 commented on server/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java:
```diff
@@ -578,7 +595,32 @@ public void testStartedShardEntrySerialization() throws Exception {
+            assertThat(deserialized.primaryTerm, equalTo(primaryTerm));
+            assertThat(deserialized.message, equalTo(message));
+            assertThat(deserialized.timestampRange, equalTo(timestampRange));
+        }
```
Good catch. Added in next push.

quux00 commented on server/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java:
```diff
@@ -578,7 +595,32 @@ public void testStartedShardEntrySerialization() throws Exception {
+        var startedShardEntry = new StartedShardEntry(shardId, allocationId, primaryTerm, message, timestampRange, eventIngestedRange);
         try (StreamInput in = serialize(startedShardEntry, version).streamInput()) {
             in.setTransportVersion(version);
             final StartedShardEntry deserialized = new StartedShardEntry(in);
```
Yes.  Added in next push.

quux00 commented on server/src/test/java/org/elasticsearch/cluster/metadata/IndexMetadataTests.java:
```diff
@@ -169,9 +185,104 @@ public void testIndexMetadataSerialization() throws IOException {
         }
     }
 
+    // MP TODO: do I need this @SuppressForbidden? What is this for?
+    @SuppressForbidden(reason = "Use IndexMetadata#getForecastedWriteLoad to ensure that the serialized value is correct")
```
Thanks. Meant to do that earlier. Good catch.

quux00 committed:
PR feedback; ShardStateActionTests asserts for eventIngestedRange; removed unneeded code comments

quux00 committed:
Added minClusterTransportVersion to XContentParserConfiguration and (some versions of) XContentParser.

This allows several more tests to pass with the new assert in IndexMetadata.Builder.eventIngestedRange
but several dozen still fail due to other classes needing to have access to minClusterTransportState that do not have it.

thecoop commented on libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentParser.java:
```diff
@@ -220,4 +220,12 @@ <T> Map<String, T> map(Supplier<Map<String, T>> mapFactory, CheckedFunction<XCon
+     * TODO: DOCUMENT ME
+     * @return
+     */
+    default Integer getMinClusterTransportVersion() {
```
This should be `TransportVersion`, not an `int`

thecoop commented:
I am super not keen for transport version to be added to xcontentparser as an int - that has a bad smell about it. Is there somewhere else that could go so its available in the right places?

quux00 committed:
Revert "Added minClusterTransportVersion to XContentParserConfiguration and (some versions of) XContentParser."

This reverts commit 0efb4ca0452f9b3d866682b39e6e526e3a69cecf.

quux00 committed:
Added eventIngestedRange Builder method that does not take TransportVersion

for use internal to IndexMetadata file

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Additional asserts around eventIngestedRange in SearchableSnapshotsIntegTests

quux00 committed:
Added additional tests or asserts around eventIngestedRange to FrozenIndexIT and FrozenIndexTests

quux00 commented:
> I am super not keen for transport version to be added to xcontentparser as an int - that has a bad smell about it. Is there somewhere else that could go so its available in the right places?

@thecoop - that commit has been reverted. After discussing with Henning we are not going to add any TransportVersion logic to the XContent config or parsers.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

henningandersen reviewed (commented):
Left a few more comments.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -2094,6 +2154,17 @@ public Builder timestampRange(IndexLongFieldRange timestampRange) {
 
+        public Builder eventIngestedRange(IndexLongFieldRange eventIngestedRange, TransportVersion minClusterTransportVersion) {
+            assert eventIngestedRange != null : "eventIngestedRange cannot be null";
+            if (minClusterTransportVersion != null
```
Yes, looks like it works now with the additional modifications you suggested. Thanks!

quux00 committed:
First round of next PR change requests

Added back the assert in public Builder.eventIngestedRange method -
 assert minClusterTransportVersion != null || eventIngestedRange == IndexLongFieldRange.UNKNOWN
now passes tests.

Changed IndexMetadataTests to use package private version of eventIngestedRange builder method that takes no TransportVersion

quux00 committed:
Added asserts around UNKNOWN for eventIngestedRange to ClusterState constructor and adjusted a few related tests

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Null check in the new assert in ClusterState to avoid some xpack test failures

quux00 committed:
Fixes to various tests or prod code to allow new ClusterState ctor assert around eventIngestedRange

Changes to:
ClusterStateDiffIT
and MetadataIndexTemplateService, which was causing several bwc tests to fail.

quux00 committed:
Additional null check in assert in ClusterState constructor to avoid test failures in TransportVersionUtilsTests

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -1652,7 +1653,11 @@ private static void validateCompositeTemplate(
-                    .put(IndexMetadata.builder(temporaryIndexName).settings(finalResolvedSettings))
+                    .put(
+                        IndexMetadata.builder(temporaryIndexName)
+                            .eventIngestedRange(IndexLongFieldRange.UNKNOWN) // necessary to pass asserts in ClusterState constructor
```
I would prefer to avoid the package private call here (while technically in the same package it is very much client code of the `IndexMetadata` class). Can we instead do:

```suggestion
                            .eventIngestedRange(IndexLongFieldRange.NO_SHARDS, state.getMinTransportVersion()) // necessary to pass asserts in ClusterState constructor
```

??

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -2131,6 +2191,29 @@ public Builder timestampRange(IndexLongFieldRange timestampRange) {
+        public Builder eventIngestedRange(IndexLongFieldRange eventIngestedRange, TransportVersion minClusterTransportVersion) {
+            assert eventIngestedRange != null : "eventIngestedRange cannot be null";
+            assert minClusterTransportVersion != null || eventIngestedRange == IndexLongFieldRange.UNKNOWN
+                : "eventIngestedRange should only be UNKNOWN if minClusterTransportVersion is null, but minClusterTransportVersion: "
```
I think this should be:

```suggestion
                : "eventIngestedRange must be UNKNOWN when minClusterTransportVersion is null, but minClusterTransportVersion: "
```

henningandersen reviewed (approved):
LGTM.

quux00 committed:
Final adjustments requested in the PR

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search/coordinator-rewrite-uses-event.ingested

quux00 committed:
Adjusted to ML_INFERENCE_GOOGLE_VERTEX_AI_EMBEDDINGS_ADDED in ClusterSerializationTests
